packages/components/src/mobile/bottom-sheet/bottom-sheet-navigation/navigation-screen.native.js

```diff
@@ -64,7 +65,7 @@ const BottomSheetNavigationScreen = ( {
 				setHeight( heightRef.current.maxHeight );
 			}
 			return () => {};
-		}, [] )
+		}, [ setHeight ] )
```

Comment:
Hey @dcalhoun , I'm trying to understand the "stale height contained in the lexical scope of the callback." :). I feel there's something for me to learn here so, can you elaborate on what happens and how the adding the `setHeight` dependency fixes/optimizes it? Thanks!

Reply:
> I'm trying to understand the "stale height contained in the lexical scope of the callback." :). I feel there's something for me to learn here so, can you elaborate on what happens and how the adding the `setHeight` dependency fixes/optimizes it? 

@hypest of course. Understanding the staleness was difficult for me as well. **TL;DR:** the memoized focus callback relied upon a memoized `setHeight` callback. The latter updated when `currentHeight` changed, the former never updated. This left the former stale with an incorrect `currentHeight`. 

First, adding `setHeight` does not optimize the callback. In fact, it de-optimizes it in the sense that the memoized focus callback now changes if the [memoized `setHeight` callback](https://github.com/WordPress/gutenberg/blob/b21c33ae1b17c56e60ce9bc6b89fe8931f20fa25/packages/components/src/mobile/bottom-sheet/bottom-sheet-navigation/navigation-container.native.js#L88-L116) changes. 

In addition to triggering the focus callback changing when `setHeight` changes, https://github.com/WordPress/gutenberg/commit/2a8d738b40a5f803045eda68e29eca8c08918af0 also limited modifications of the bottom sheet height to scenarios where [the height value actually changed](https://github.com/WordPress/gutenberg/commit/2a8d738b40a5f803045eda68e29eca8c08918af0#diff-4ff314b0b4c42d8ed1ff31f1b7d333c37bdfd0ab344e9d0f0d29c41c5cc004e0R101). This limitation is the _actual bug fix_, as it prevents registering a `LayoutAnimation` that will be left unconsumed by setting the height to same value as the current height. 

To enable the aforementioned bug fix to work, the `height` and `currentHeight` values must be up-to-date, otherwise the comparison will be stale and incorrect. The code was layering memoized callbacks where the base callback (`setHeight`) changed when `currentHeight` updated, but the layered callback (focus callback) did not change as `setHeight` was not declared as a memoization dependency. This resulted in `currentHeight` becoming stale within the the lexical scope of the focus callback. 

I added [two tests](https://github.com/WordPress/gutenberg/blob/b21c33ae1b17c56e60ce9bc6b89fe8931f20fa25/packages/components/src/mobile/bottom-sheet/bottom-sheet-navigation/test/navigation-container.native.js#L70-L162) to cover the original bug where a `LayoutAnimation` was left unconsumed. 

Please let me know if there is more I can explain on this subject.

Reply:
Ah, that helps me understand, thanks for elaborating David! And, good catch I'd say 😄

Reply:
Related to this topic, I stumbled upon https://github.com/WordPress/gutenberg/pull/26773 today. This change reintroduces this bug. 

The aforementioned PR leveraged `useCallback`'s dependency array as a "fix" for business logic. Doing so is a brittle approach and explicitly discouraged in the React.js documentation, as React may one day not guarantee the memoized function will not change. 

>

Reply:
Related to this topic, I stumbled upon https://github.com/WordPress/gutenberg/pull/26773 today. This change reintroduces the Android back button bug. 

The aforementioned PR leveraged `useCallback`'s dependency array as a "fix" for business logic. Doing so is a brittle approach and [explicitly discouraged](https://reactjs.org/docs/hooks-reference.html#usecallback) in the React.js documentation, as React may one day not guarantee the memoized function will not change. 

> **You may rely on useMemo as a performance optimization, not as a semantic guarantee.** In the future, React may choose to “forget” some previously memoized values and recalculate them on next render, e.g. to free memory for offscreen components. Write your code so that it still works without `useMemo` — and then add it to optimize performance.

I plan to look into an alternative fix for the Android back button issue.